Application passwords reauthentication flow

crazytonyli opened the pull request:
## Description

Considering users have very easy access to application passwords on wp-admin, and there is no super obvious indicator on wp-admin to show that the official WordPress/Jetpack app is currently using one application password, I _presume_ users may accidentally revoke the one used by the app. This PR introduces a reauthentication flow when the app notices that the application password becomes invalid. The flow is similar to where WP.com users disconnect the app from their "Connected Application" webpage on WP.com.

## Testing instructions

Here is a video demo:

https://github.com/user-attachments/assets/8a38c6f3-1c8a-437e-a172-1a1b79264856.mov

<!--

Consider testing the following *yourself* before requesting a review:

- Compatibility with WordPress.com sites and self-hosted Jetpack sites
- Both portrait and landscape orientations
- Light and dark modes
- Dynamic font sizes: larger, smaller and bold text
- High contrast mode
- Screen reader experience (e.g., VoiceOver)
- Languages with large words or with letters/accents not frequently used in English
- Right-to-left languages layout
- Multi-tasking: split view and slide over

-->

crazytonyli committed:
Update wordpress-rs

crazytonyli committed:
Do not overwrite account passwords with application passwords

crazytonyli committed:
Prompt for re-authentication when the application password becomes invalid

dangermattic commented:
<!--
  0 Errors
  2 Warnings: <code>Modules/Package.swift</c...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/310">26.0</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

crazytonyli committed:
Fix unit tests compiling issues

kean commented on WordPress/Classes/Plugins/Views/InstalledPluginsListView.swift:
```diff
@@ -54,6 +54,9 @@ struct InstalledPluginsListView: View {
                 AddNewPluginView(service: viewModel.service)
             }
         }
+        .onApplicationPasswordUpdate {
```
(nit) I think it's an over-kill to add a ViewModifier just to listen to a notification.

crazytonyli replied:
I agree. An extension function is enough here.

kean reviewed (approved):
Testing it, and it worked OK.

Two observations unrelated to the PR:

1. It's weird that you can add the same site twice (per account)

<img width="320" alt="Screenshot 2025-05-07 at 10 32 06 AM" src="https://github.com/user-attachments/assets/30bb5e29-91ae-4e2a-9107-5e3d15eaa980" />

I'd suggest making it impossible. If you are adding the same site – throw an error and ask you to logout if you need to change the account.

2. The current application password flow has all the global navigation in it, so it's easy to mess it up. It doesn't seem to be designed for this use case.

crazytonyli commented:
> The current application password flow has all the global navigation in it,

Can you elaborate on this? Do you mean how the API client instance uses notification to call into the app-wide code, which then calls the `SelfHostedSiteAuthenticator` to reauthenticate?

crazytonyli committed:
Delete `ApplicationPasswordUpdatedViewModifier`

crazytonyli committed:
Update an unit test

Blog.password was overwritten to the new application password. We no
longer do that anymore.

kean commented:
> Can you elaborate on this? Do you mean how the API client instance uses notification to call into the app-wide code, which then calls the SelfHostedSiteAuthenticator to reauthenticate?

I just mean this:

<img width="320" alt="Screenshot 2025-05-07 at 8 09 33 PM" src="https://github.com/user-attachments/assets/f84c985b-d2a3-4491-97e2-b7283036e3ec" />

I also ran into some issue when testing adding the two accounts for the same site – the web view remembers the previous user, so it's hard to manage. I think we should restrict this scenario.
